File: example/rust/src/bin/4_pos_vel_control.rs
use motor_vendor_damiao::ControlMode;
use rebotarm_control_rt_rust_examples::common::{
    default_vlim, has_flag, parse_floats, parse_port, parse_rate, prompt, sleep_to_rate, B601Arm,
    ALL_DOF, B601_JOINTS, DEFAULT_RATE_HZ,
};
use std::env;
use std::error::Error;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Instant;

fn main() -> Result<(), Box<dyn Error>> {
    let args: Vec<String> = env::args().skip(1).collect();
    if has_flag(&args, "--help") || has_flag(&args, "-h") {
        println!("Usage: cargo run --bin 4_pos_vel_control -- --port /dev/ttyACM0 --rate 150");
        return Ok(());
    }

    let rate = parse_rate(&args, DEFAULT_RATE_HZ);
    let arm = B601Arm::open(&parse_port(&args))?;
    arm.enable()?;
    arm.ensure_all_mode(ControlMode::PosVel);

    let target = Arc::new(Mutex::new(arm.positions_or_zero()));
    let vlim = Arc::new(Mutex::new(default_vlim()));
    let running = Arc::new(AtomicBool::new(true));
    let motors = arm.motors.clone();
    let target_rt = Arc::clone(&target);
    let vlim_rt = Arc::clone(&vlim);
    let running_rt = Arc::clone(&running);

    let handle = thread::spawn(move || {
        while running_rt.load(Ordering::Relaxed) {
            let tick = Instant::now();
            let target = target_rt.lock().map(|v| v.clone()).unwrap_or_default();
            let vlim = vlim_rt.lock().map(|v| v.clone()).unwrap_or_default();
            for idx in 0..motors.len() {
                let _ = motors[idx].send_cmd_pos_vel(
                    target.get(idx).copied().unwrap_or(0.0),
                    vlim.get(idx).copied().unwrap_or(B601_JOINTS[idx].vlim),
                );
            }
            sleep_to_rate(tick, rate);
        }
    });

    println!("POS_VEL loop started at {rate:.1} Hz.");
    println!("Input: q1 ... q6 [gripper] [vlim], state, or q. Angles are degrees.");
    loop {
        let Some(line) = prompt("> ")? else {
            break;
        };
        if line.is_empty() {
            continue;
        }
        if matches!(line.as_str(), "q" | "quit" | "exit") {
            break;
        }
        if line == "state" {
            arm.print_state();
            continue;
        }
        let values = parse_floats(&line)?;
        if values.len() < 6 {
            println!("need at least 6 joint angles");
            continue;
        }
        {
            let mut target = target.lock().expect("target lock");
            for idx in 0..ALL_DOF.min(values.len()) {
                target[idx] = values[idx].to_radians() as f32;
            }
        }
        if values.len() >= ALL_DOF + 1 {
            vlim.lock().expect("vlim lock").fill(values[ALL_DOF] as f32);
        }
    }

    running.store(false, Ordering::Relaxed);
    let _ = handle.join();
    arm.close();
    Ok(())
}

File: example/cpp/src/4_pos_vel_control.cpp
#include "example_common.hpp"

#include <atomic>
#include <exception>
#include <iostream>
#include <mutex>
#include <thread>

int main(int argc, char** argv) {
  try {
    if (example::has_flag(argc, argv, "--help") || example::has_flag(argc, argv, "-h")) {
      std::cout << "Usage: ./4_pos_vel_control --port /dev/ttyACM0 --rate 150\n";
      return 0;
    }

    const double rate = example::parse_rate(argc, argv, example::kDefaultRateHz);
    auto arm = example::B601Arm::open(example::parse_port(argc, argv));
    arm.enable();
    arm.ensure_all_mode(motorbridge::Mode::POS_VEL);

    std::mutex mutex;
    std::vector<float> target = arm.positions_or_zero();
    std::vector<float> vlim = example::default_vlim();
    std::atomic_bool running{true};

    std::thread loop([&]() {
      while (running.load(std::memory_order_relaxed)) {
        const auto tick = std::chrono::steady_clock::now();
        std::vector<float> target_copy;
        std::vector<float> vlim_copy;
        {
          std::lock_guard<std::mutex> lock(mutex);
          target_copy = target;
          vlim_copy = vlim;
        }
        for (int i = 0; i < static_cast<int>(arm.motors.size()); ++i) {
          try {
            arm.motors[i].send_pos_vel(target_copy[i], vlim_copy[i]);
          } catch (...) {
          }
        }
        example::sleep_to_rate(tick, rate);
      }
    });

    std::cout << "POS_VEL loop started at " << rate << " Hz.\n";
    std::cout << "Input: q1 ... q6 [gripper] [vlim], state, or q. Angles are degrees.\n";
    while (true) {
      const auto line_opt = example::prompt("> ");
      if (!line_opt) break;
      const std::string line = *line_opt;
      if (line == "q" || line == "quit" || line == "exit") break;
      if (line == "state") {
        arm.print_state();
        continue;
      }
      if (line.empty()) continue;
      const auto values = example::parse_floats(line);
      if (values.size() < 6) {
        std::cout << "need at least 6 joint angles\n";
        continue;
      }
      std::lock_guard<std::mutex> lock(mutex);
      for (int i = 0; i < std::min(example::kAllDof, static_cast<int>(values.size())); ++i) {
        target[i] = example::deg_to_rad_f32(values[i]);
      }
      if (values.size() >= example::kAllDof + 1) {
        std::fill(vlim.begin(), vlim.end(), static_cast<float>(values[example::kAllDof]));
      }
    }

    running.store(false, std::memory_order_relaxed);
    loop.join();
    arm.close();
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "error: " << e.what() << "\n";
    return 1;
  }
}
